calicoctl/commands/label.go

```diff
@@ -0,0 +1,142 @@
+  calicoctl label nodes node1 cluster= --overwrite
+
+Options:
+  -h --help                    Show this screen.
+  -c --config=<CONFIG>         Path to the file containing connection
+                               configuration in YAML or JSON format.
+                               [default: ` + constants.DefaultConfigPath + `]
+  -n --namespace=<NS>          Namespace of the resource.
+                               Only applicable to NetworkPolicy and WorkloadEndpoint.
+                               Uses the default namespace if not specified.
+  --overwrite                  If true, overwrite the value when the key is already
+                               present in labels. Otherwise reports error when the
+                               labeled resource already have the key in its labels.
+
+Description:
+  The label command is used to add or update a label on a resource. Resource types
+  that can be labeled are:
+
+    * bgpConfiguration
+    * bgpPeer
+    * felixConfiguration
+    * globalNetworkPolicy
+    * globalNetworkSet
+    * hostEndpoint
+    * ipPool
+    * networkPolicy
+    * node
+    * profile
+    * workloadEndpoint
+
+  The resource type is case insensitive and may be pluralized.
+
+  Attempting to label resources that are not exist will get an error.
+
+  When label a resource on an existing key in its labels:
+  - gets an error if option --overwrite is not provided.
+  - value of the key updates to specified value if option --overwrite is provided. 
+  `
+
+	parsedArgs, err := docopt.Parse(doc, args, true, "", false, false)
+	if err != nil {
+		fmt.Printf("Invalid option: 'calicoctl %s'. Use flag '--help' to read about a specific subcommand.\n", strings.Join(args, " "))
+		os.Exit(1)
+	}
+	if len(parsedArgs) == 0 {
+		return
+	}
+
+	fmt.Printf("parse args: %+v\n", parsedArgs)
+	// get results.
+	kind := parsedArgs["<KIND>"].(string)
+	name := parsedArgs["<NAME>"].(string)
+	results := executeConfigCommand(parsedArgs, actionGetOrList)
+	if results.fileInvalid {
+		fmt.Printf("Failed to execute command: %v\n", results.err)
+		os.Exit(1)
+	} else if results.err != nil {
+		fmt.Printf("failed to get %s %s, error %v\n",
+			kind, name, results.err)
+		os.Exit(1)
+	} else if len(results.resources) == 0 {
+		fmt.Printf("%s %s not found\n", kind, name)
+		os.Exit(1)
+	}
+	kv := strings.Split(parsedArgs["<key>=<value>"].(string), "=")
+	if len(kv) != 2 {
+		fmt.Printf("invalid label %s\n", parsedArgs["<key>=<value>"])
+		os.Exit(1)
+	}
+	key := kv[0]
+	value := kv[1]
+
+	resource := results.resources[0].(resourcemgr.ResourceObject)
+	labels := resource.GetObjectMeta().GetLabels()
+	overwrite := parsedArgs["--overwrite"].(bool)
+	overwritten := false
+	client := results.client
+
+	oldValue, ok := labels[key]
+	if ok {
+		if overwrite || value == oldValue {
+			labels[key] = value
+			overwritten = true
+		} else {
+			fmt.Printf("failed to update label of %s %s, key %s is already present\n",
```

Comment:
Would be nice to log out something like "use --overwrite to set a new value" to indicate the user needs to set an additional flag.

Reply:
moved the parsing to the begging of func.